scheduler/src/cook/kubernetes/api.clj

```diff
@@ -1836,6 +1837,43 @@
               (log/info "In" compute-cluster-name "compute cluster, pod" pod-name "was already deleted")
               (throw e))))))))
 
+(defn delete-collect-results-finalizer
+  "Delete the finalizer that collects the results of a pod completion state once the pod has a
+  deletion timestamp"
+  [{:keys [compute-cluster-name api-client name]} pod-name ^V1Pod pod]
+  (let [^V1ObjectMeta pod-metadata (some-> pod .getMetadata)
+        pod-deletion-timestamp (some-> pod .getMetadata .getDeletionTimestamp)
+        ^List finalizers (some-> pod-metadata .getFinalizers)]
+    (when (and finalizers (> (.size finalizers) 1))
+      ; Finalizers are stored in a list and deleted by index. If there's more than one,
+      ; cook can race with whatever deletes the other one, or even delete both (see race discussed below)
+      (log/error "In compute-cluster" name ", deleting finalizer for pod" pod-name "Cook's deletion of more than one finalizer on a pod is racy and not supported. Finalizers are" finalizers))
+    (when (and pod-deletion-timestamp (some #(= FinalizerHelper/collectResultsFinalizer %) finalizers))
+      (log/info "In compute-cluster" name ", deleting finalizer for pod" pod-name)
+      (timers/time! (metrics/timer "delete-finalizer" compute-cluster-name)
+        (try
+          (FinalizerHelper/removeFinalizer api-client pod)
+          (catch ApiException e
+            (let [code (.getCode e)]
+              ; There can be races here. Whenever we hit the state machine, we will run this code
+              ; and may try to delete the finalizer twice. The second invocation can see the pod
+              ; already deleted or it can see an unprocesssable entity as we try to delete a finalizer
+              ; that's already gone.
+              (when (= 404 code)
+                (meters/mark! (metrics/meter "delete-finalizer-expected-errors" compute-cluster-name))
+                (log/info "In compute-cluster" name ", Not Found error deleting finalizer for pod" pod-name))
+              (when (= 422 code)
+                (meters/mark! (metrics/meter "delete-finalizer-expected-errors" compute-cluster-name))
+                (log/info "In compute-cluster" name ", Unprocessable Entity error deleting finalizer for pod" pod-name))
+              (when-not (contains? #{404 422} code)
+                (meters/mark! (metrics/meter "delete-finalizer-unexpected-errors" compute-cluster-name))
+                (log/error e "In compute-cluster" name ", error deleting finalizer for pod" pod-name))))
+          (catch Exception e
+            (meters/mark! (metrics/meter "delete-finalizer-unexpected-errors" compute-cluster-name))
+            (log/error e "In compute-cluster" name ", error deleting finalizer for pod" pod-name)))))))
```

Comment:
To make these logs consistent with the existing logs, can we make them be `"In" name "compute cluster, ..."`

Reply:
Sure. FYI: About half of the log lines are `"In compute-cluster" name`. The other half are `"In" name "compute cluster`.

Reply:
Hmm, what am I missing? I'm not getting about half:

```
$ git grep 'In compute-cluster' | wc -l
4

$ git grep 'In compute cluster' | wc -l
21

$ git grep '"compute cluster, ' | wc -l
50
```

Reply:
$ git grep 'In compute cluster' | wc
     21     243    2702
$ git grep 'In *" *name *" *compute cluster' | wc
     18     206    2264
$ git grep 'In compute-cluster'  | wc
      4      52     539

Reply:
```
$ git grep 'In.*".*name.*".*compute cluster' | wc -l
48

$ git grep 'In.*".*name.*".*compute cluster'
scheduler/src/cook/kubernetes/api.clj:      (log/info "In" compute-cluster-name "compute cluster, listing pods for all namespaces"
scheduler/src/cook/kubernetes/api.clj:    (log/info "In" compute-cluster-name "compute cluster, pod watch processing pods"
scheduler/src/cook/kubernetes/api.clj:                                 (log/info "In" compute-cluster-name "compute cluster, pod watch doing callback"
scheduler/src/cook/kubernetes/api.clj:          (log/info "In" compute-cluster-name "compute cluster, handling pod watch updates")
scheduler/src/cook/kubernetes/api.clj:                (log/info e "In" compute-cluster-name "compute cluster, pod watch timed out")
scheduler/src/cook/kubernetes/api.clj:                (log/error e "In" compute-cluster-name "compute cluster, error during pod watch"))))
scheduler/src/cook/kubernetes/api.clj:  (log/info "In" compute-cluster-name "compute cluster, initializing pod watch")
scheduler/src/cook/kubernetes/api.clj:                  (log/info "In" compute-cluster-name "compute cluster, initializing pod watch helper")
scheduler/src/cook/kubernetes/api.clj:    (log/info "In" compute-cluster-name "compute cluster, node watch processing nodes"
scheduler/src/cook/kubernetes/api.clj:      (log/info "In" compute-cluster-name "compute cluster, node watch doing callback"
scheduler/src/cook/kubernetes/api.clj:          (log/info "In" compute-cluster-name "compute cluster, handling node watch updates")
scheduler/src/cook/kubernetes/api.clj:  (log/info "In" compute-cluster-name "compute cluster, initializing node watch")
scheduler/src/cook/kubernetes/api.clj:                  (log/info "In" compute-cluster-name "compute cluster, initializing node watch helper")
scheduler/src/cook/kubernetes/api.clj:        (log/info "In" compute-cluster-name "compute cluster, handling event watch updates")
scheduler/src/cook/kubernetes/api.clj:              (log/info e "In" compute-cluster-name "compute cluster, event watch timed out")
scheduler/src/cook/kubernetes/api.clj:              (log/error e "In" compute-cluster-name "compute cluster, error during event watch"))))
scheduler/src/cook/kubernetes/api.clj:  (log/info "In" compute-cluster-name "compute cluster, initializing event watch")
scheduler/src/cook/kubernetes/api.clj:                  (log/info "In" compute-cluster-name "compute cluster, initializing event watch helper")
scheduler/src/cook/kubernetes/api.clj:        "In" name "compute cluster, encountered node with GPU taint but no allocatable GPUs"
scheduler/src/cook/kubernetes/api.clj:          (log/info "In" compute-cluster-name "compute cluster, caught the https://github.com/kubernetes-client/java/issues/252 exception deleting" pod-name))
scheduler/src/cook/kubernetes/api.clj:              (log/info "In" compute-cluster-name "compute cluster, pod" pod-name "was already deleted")
scheduler/src/cook/kubernetes/api.clj:        (log/info "In" compute-cluster-name "compute cluster, launching pod with name" pod-name "in namespace" namespace ":" (.serialize json pod))
scheduler/src/cook/kubernetes/api.clj:              (log/info e "In" compute-cluster-name "compute cluster, error submitting pod"
scheduler/src/cook/kubernetes/compute_cluster.clj:    (log/info "In" compute-cluster-name "compute cluster, generating offers"
scheduler/src/cook/kubernetes/compute_cluster.clj:  (log/info "In" name "compute cluster, starting taskid scan")
scheduler/src/cook/kubernetes/compute_cluster.clj:    (log/info "In" name "compute cluster, doing taskid scan. Visiting" (count taskids) "taskids")
scheduler/src/cook/kubernetes/compute_cluster.clj:        (log/info "In" name "compute cluster, doing scan of" taskid)
scheduler/src/cook/kubernetes/compute_cluster.clj:        (log/error e "In" name "compute cluster, encountered exception launching task"
scheduler/src/cook/kubernetes/compute_cluster.clj:      (log/info "In" pool-name "pool, launching tasks for" name "compute cluster"
scheduler/src/cook/kubernetes/compute_cluster.clj:        (log/error "In" name "compute cluster, attempting to delete task, with ID" task-id
scheduler/src/cook/kubernetes/compute_cluster.clj:        (log/info "In" name "compute cluster, skipping generating offers for pool" pool-name
scheduler/src/cook/kubernetes/compute_cluster.clj:            (log/info "In" name "compute cluster, not looking for offers for pool" pool-name
scheduler/src/cook/kubernetes/compute_cluster.clj:              (log/info "In" name "compute cluster, looking for offers for pool" pool-name)
scheduler/src/cook/kubernetes/compute_cluster.clj:                (log/info "In" name "compute cluster, generated offers for pool"
scheduler/src/cook/kubernetes/compute_cluster.clj:      (log/warn "In" name "compute cluster, ignoring request to autoscale in pool" pool-name
scheduler/src/cook/kubernetes/compute_cluster.clj:                (str "In " name " compute cluster, request to autoscale despite invalid / missing config"))
scheduler/src/cook/kubernetes/compute_cluster.clj:            (log/warn "In" name "compute cluster, total pods are maxed out"
scheduler/src/cook/kubernetes/compute_cluster.clj:            (log/warn "In" name "compute cluster, nodes are maxed out"
scheduler/src/cook/kubernetes/compute_cluster.clj:            (log/warn "In" name "compute cluster, synthetic pods are maxed out"
scheduler/src/cook/kubernetes/compute_cluster.clj:              (log/warn "In" name "compute cluster, cannot launch more synthetic pods"
scheduler/src/cook/kubernetes/compute_cluster.clj:                (log/info "In" name "compute cluster, launching" num-synthetic-pods-to-launch
scheduler/src/cook/kubernetes/compute_cluster.clj:          (log/error e "In" name "compute cluster, encountered error launching synthetic pod(s) in"
scheduler/src/cook/kubernetes/compute_cluster.clj:      (throw (ex-info (str "In " compute-cluster-name " compute cluster, invalid synthetic pods config")
scheduler/src/cook/kubernetes/controller.clj:      (log/info "In" name "compute cluster, processing pod" pod-name ";"
scheduler/src/cook/mesos/mesos_compute_cluster.clj:          (log/info "In" compute-cluster-name "compute cluster, mesomatic scheduler join unblocked; we appear to have lost driver connection")
scheduler/src/cook/mesos/mesos_compute_cluster.clj:            (log/warn e "In" compute-cluster-name "compute cluster, lost mesos driver with exception")
scheduler/src/cook/mesos/mesos_compute_cluster.clj:    (log/info "In" compute-cluster-name "compute cluster, looking for offers for pool" pool-name)
scheduler/src/cook/scheduler/scheduler.clj:                         _ (log/info "In" pool-name "pool, launching matched tasks for" compute-cluster-name "compute cluster")
```

Reply:
That's a better grep than mine. So about 30%, not 50%